packages/story-editor/src/components/panels/document/taxonomies/HierarchicalTermSelector.js

```diff
@@ -95,33 +99,105 @@ const AddNewCategoryButton = styled(Button).attrs({
 `;
 
 function HierarchicalTermSelector({ taxonomy }) {
+  const { createTerm, selectedSlugs, setSelectedTaxonomySlugs, termCache } =
+    useTaxonomy(
+      ({
+        state: { selectedSlugs, termCache },
+        actions: { createTerm, setSelectedTaxonomySlugs },
+      }) => ({
+        createTerm,
+        selectedSlugs,
+        setSelectedTaxonomySlugs,
+        termCache,
+      })
+    );
+
+  // Categories to
+  const categories = useMemo(() => {
+    if (termCache[taxonomy.rest_base]) {
+      return Object.values(termCache[taxonomy.rest_base]).map((category) => {
+        const formattedCategory = { ...category };
+        formattedCategory.value = formattedCategory.id;
+        formattedCategory.label = formattedCategory.name;
+        formattedCategory.checked = selectedSlugs[taxonomy.rest_base]?.includes(
+          category.slug
+        );
+
+        return formattedCategory;
+      });
+    }
+
+    return [];
+  }, [selectedSlugs, taxonomy, termCache]);
+
   const [showAddNewCategory, setShowAddNewCategory] = useState(false);
   const [newCategoryName, setNewCategoryName] = useState('');
+  const [selectedParent, setSelectedParent] = useState();
   const dropdownId = useMemo(uuidv4, []);
 
-  const handleShowAddNewCategoryClick = useCallback(() => {
-    setShowAddNewCategory(true);
+  const resetInputs = useCallback(() => {
+    setNewCategoryName('');
+    setSelectedParent();
   }, []);
 
+  const handleClickCategory = useCallback(
+    (_evt, { id, checked }) => {
+      const term = categories.find((category) => category.id === id);
+
+      // find the already selected slugs + update those.
+      setSelectedTaxonomySlugs(taxonomy, (currentTerms = []) => {
+        const index = currentTerms.findIndex((slug) => slug === term.slug);
+        // add if term doesn't exist
+        if (checked && index === -1) {
+          return [...currentTerms, term.slug];
+        }
+
+        // remove if term exists
+        if (!checked && index > -1) {
+          return [
+            ...currentTerms.slice(0, index),
+            ...currentTerms.slice(index + 1),
+          ];
+        }
+
+        return currentTerms;
+      });
+    },
+    [categories, setSelectedTaxonomySlugs, taxonomy]
+  );
+
+  const handleToggleNewCategory = useCallback(() => {
+    setShowAddNewCategory((currentValue) => !currentValue);
+    resetInputs();
+  }, [resetInputs]);
+
   const handleChangeNewCategoryName = useCallback((evt) => {
     setNewCategoryName(evt.target.value);
   }, []);
 
   const handleAddCategory = useCallback(() => {
+    createTerm(taxonomy, newCategoryName, selectedParent);
     setShowAddNewCategory(false);
-  }, []);
+    resetInputs();
+  }, [createTerm, newCategoryName, resetInputs, selectedParent, taxonomy]);
```

Comment:
Looks like Gutenberg adds the new category as checked when initially created. Do we want to add that in here? or was that an intentional UX deviation?

Reply:
Added a new issue for this! https://github.com/google/web-stories-wp/issues/9058